src/test/java/fi/aalto/cs/apluscourses/presentation/module/ModuleListElementViewModelTest.java

```diff
@@ -0,0 +1,83 @@
+import java.net.URL;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.junit.Test;
+
+public class ModuleListElementViewModelTest {
+
+  @Test
+  public void testStateChanged() throws MalformedURLException {
+    AtomicBoolean isOnChangedCalled = new AtomicBoolean(false);
+    Module module = new Module("testModule", new URL("https://example.com/1"));
+    ModuleListElementViewModel moduleViewModel = new ModuleListElementViewModel(module) {
+      @Override
+      public void onChanged() {
+        // For some weird reason, Mockito's spy-verify could not recognize this method call.
+        isOnChangedCalled.set(true);
+      }
+    };
+    module.stateMonitor.set(Module.FETCHING);
+    assertTrue(isOnChangedCalled.get());
+    assertNotNull(moduleViewModel); // prevent weak reference from expiring
+  }
+
+  @Test
+  public void testNameAndUrl() throws MalformedURLException {
+    String name = "Wanda";
+    String url = "https://example.com/wanda";
+    Module module = new Module(name, new URL(url));
+    ModuleListElementViewModel moduleViewModel = new ModuleListElementViewModel(module);
+    assertEquals("getName() should return module's name",
+        name, moduleViewModel.getName());
+    assertEquals("getUrl() should return module's URL",
+        url, moduleViewModel.getUrl());
+  }
+
+  @Test
+  public void testStatus() throws MalformedURLException {
+    Module module = new Module("testStatusModule", new URL("https://example.com/2"));
+    ModuleListElementViewModel moduleViewModel = new ModuleListElementViewModel(module);
+
+    float delta = 0.001f;
+    assertEquals("Not installed", moduleViewModel.getStatus());
+    assertEquals(TextAttribute.WEIGHT_REGULAR, moduleViewModel.getFontWeight(), delta);
+
+    module.stateMonitor.set(Module.FETCHING);
+    assertEquals("Downloading...", moduleViewModel.getStatus());
+    assertEquals(TextAttribute.WEIGHT_REGULAR, moduleViewModel.getFontWeight(), delta);
+
+    module.stateMonitor.set(Module.FETCHED);
+    assertEquals("Not installed", moduleViewModel.getStatus());
+    assertEquals(TextAttribute.WEIGHT_REGULAR, moduleViewModel.getFontWeight(), delta);
+
+    module.stateMonitor.set(Module.LOADING);
+    assertEquals("Installing...", moduleViewModel.getStatus());
+    assertEquals(TextAttribute.WEIGHT_REGULAR, moduleViewModel.getFontWeight(), delta);
+
+    module.stateMonitor.set(Module.LOADED);
+    assertEquals("Loaded", moduleViewModel.getStatus());
+    assertEquals(TextAttribute.WEIGHT_BOLD, moduleViewModel.getFontWeight(), delta);
+
+    module.stateMonitor.set(Module.WAITING_FOR_DEPS);
+    assertEquals("Waiting for dependencies...", moduleViewModel.getStatus());
+    assertEquals(TextAttribute.WEIGHT_BOLD, moduleViewModel.getFontWeight(), delta);
+
+    module.stateMonitor.set(Module.INSTALLED);
+    assertEquals("Installed", moduleViewModel.getStatus());
+    assertEquals(TextAttribute.WEIGHT_BOLD, moduleViewModel.getFontWeight(), delta);
```

Comment:
Another benefit of Presentation Model (= view model) pattern: being able to unit test presentation logic (such as here) without dealing with UI classes.